api/grpcserver/events.go

```diff
@@ -10,25 +10,43 @@ import (
 
 const subscriptionChanBufSize = 1 << 16
 
-func consumeEvents(ctx context.Context, in <-chan interface{}) <-chan interface{} {
-	out := make(chan interface{}, subscriptionChanBufSize)
+// Errors for cases with a full event buffer.
+var (
+	errTxBufferFull          = "tx buffer is full"
+	errLayerBufferFull       = "layer buffer is full"
+	errAccountBufferFull     = "account buffer is full"
+	errRewardsBufferFull     = "rewards buffer is full"
+	errReceiptsBufferFull    = "receipts buffer is full"
+	errActivationsBufferFull = "activations buffer is full"
+	errStatusBufferFull      = "status buffer is full"
+	errErrorsBufferFull      = "errors buffer is full"
+)
+
+func consumeEvents(ctx context.Context, in <-chan interface{}) (out <-chan interface{}, bufFull <-chan struct{}) {
+	outCh := make(chan interface{}, subscriptionChanBufSize)
+	bufFullCh := make(chan struct{})
 
 	go func() {
-		defer close(out)
+		defer func() {
+			close(outCh)
+			for range in {
+			}
```

Comment:
rm?

Reply:
i assume this is added to drain the subscription channel until we close the subscription. otherwise we will block the emitter, which is the exact thing we want to prevent. can we close the subscription instead when the buffer is full?